File: src/swift/3_StructuralPatterns/4_Decorator/MultipleInheritance.swift
protocol ICanFly
{
  func fly()
}

protocol ICanCrawl
{
  func crawl()
}

class Bird: ICanFly
{
  func fly(){}
}

class Lizard: ICanCrawl
{
  func crawl(){}
}

class Dragon: ICanFly, ICanCrawl
{
  private let bird = Bird()
  private let lizard = Lizard()

  func fly() {
    bird.fly() // delegation
  }

  func crawl() {
    lizard.crawl()
  }
}

File: src/csharp/3_StructuralPatterns/4_Decorator/MultipleInheritance.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DotNetDesignPatternDemos.Structural.Decorator
{
  public class Bird
  {
    public void Fly()
    {
      
    }
  }

  public class Lizard
  {
    public void Crawl()
    {
      
    }
  }

  public class Dragon // no multiple inheritance
  {
    private Bird bird;
    private Lizard lizard;

    public Dragon(Bird bird, Lizard lizard)
    {
      this.bird = bird ?? throw new ArgumentNullException(paramName: nameof(bird));
      this.lizard = lizard ?? throw new ArgumentNullException(paramName: nameof(lizard));
    }

    public void Crawl()
    {
      lizard.Crawl();
    }

    public void Fly()
    {
      bird.Fly();
    }
  }
}
